port/cpl_minizip_zip.cpp

```diff
@@ -2141,6 +2272,297 @@ CPLErr CPLCloseFileInZip(void *hZip)
+            return CE_Failure;
+        }
+        sozip_index.resize(32);
+        uint32_t nVal32;
+        // Version
+        nVal32 = CPL_LSBWORD32(1);
+        memcpy(sozip_index.data(), &nVal32, sizeof(nVal32));
+        // Extra reserved space after 32 bytes of header
+        nVal32 = CPL_LSBWORD32(0);
+        memcpy(sozip_index.data() + 4, &nVal32, sizeof(nVal32));
+        // Chunksize
+        nVal32 = CPL_LSBWORD32(nChunkSize);
+        memcpy(sozip_index.data() + 8, &nVal32, sizeof(nVal32));
+        // SOZIPIndexEltSize
+        nVal32 = CPL_LSBWORD32(static_cast<uint32_t>(nOffsetSize));
+        memcpy(sozip_index.data() + 12, &nVal32, sizeof(nVal32));
+        // Uncompressed size
+        uint64_t nVal64 = nUncompressedSize;
+        CPL_LSBPTR64(&nVal64);
+        memcpy(sozip_index.data() + 16, &nVal64, sizeof(nVal64));
+        zi->sozip_index = &sozip_index;
+
+        zi->nChunkSize = nChunkSize;
+
+        const char *pszThreads = CSLFetchNameValue(papszOptions, "NUM_THREADS");
+        if (pszThreads == nullptr || EQUAL(pszThreads, "ALL_CPUS"))
+            zi->nThreads = CPLGetNumCPUs();
+        else
+            zi->nThreads = atoi(pszThreads);
+        zi->nThreads = std::max(1, std::min(128, zi->nThreads));
+    }
+
+    aosNewsOptions.SetNameValue("ZIP64",
+                                nUncompressedSize > 0xFFFFFFFFU ? "YES" : "NO");
+
+    if (pszInputFilename != nullptr &&
+        aosNewsOptions.FetchNameValue("TIMESTAMP") == nullptr)
+    {
+        VSIStatBufL sStat;
+        if (VSIStatL(pszInputFilename, &sStat) == 0 && sStat.st_mtime != 0)
+        {
+            aosNewsOptions.SetNameValue(
+                "TIMESTAMP",
+                CPLSPrintf(CPL_FRMT_GIB, static_cast<GIntBig>(sStat.st_mtime)));
+        }
+    }
+
+    if (CPLCreateFileInZip(hZip, pszArchiveFilename, aosNewsOptions.List()) !=
+        CE_None)
+    {
+        zi->sozip_index = nullptr;
+        zi->nChunkSize = 0;
+        zi->nThreads = 0;
+        return CE_Failure;
+    }
+    zi->nChunkSize = 0;
+    zi->nThreads = 0;
+
+    std::vector<GByte> abyChunk(1024 * 1024);
+    vsi_l_offset nOffset = 0;
+    while (true)
+    {
+        const int nRead = static_cast<int>(
+            VSIFReadL(abyChunk.data(), 1, abyChunk.size(), fpInput));
+        if (nRead > 0 &&
+            CPLWriteFileInZip(hZip, abyChunk.data(), nRead) != CE_None)
+        {
+            CPLCloseFileInZip(hZip);
+            zi->sozip_index = nullptr;
+            return CE_Failure;
+        }
+        nOffset += nRead;
+        if (pProgressFunc &&
+            !pProgressFunc(nUncompressedSize == 0
+                               ? 1.0
+                               : double(nOffset) / nUncompressedSize,
+                           nullptr, pProgressData))
+        {
+            CPLCloseFileInZip(hZip);
+            zi->sozip_index = nullptr;
+            return CE_Failure;
+        }
+        if (nRead < 1024 * 1024)
```

Comment:
Shouldn't this check be against abyChunk.size() ?

Reply:
done